### PR TITLE
feat: add function in SafehillApi.kt to get users from phone number hash (SAF-378)

### PR DESCRIPTION
**Pull request by DikenMaharjan**

### CODE REVIEW

File: src/main/kotlin/com/safehill/kclient/api/dtos/UserHashedPhoneNumbersDTO.kt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.safehill.kclient.api.dtos
+
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class UserHashedPhoneNumbersDTO(
```
Comment (gennaro-safehill) [inline src/main/kotlin/com/safehill/kclient/api/dtos/UserHashedPhoneNumbersDTO.kt]: Name this as it's named on Safehill-Server. `UserPhoneNumbersDTO`

File: src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt [hunk 2]
```diff
@@ -36,6 +44,39 @@ data class SHRemoteUser(
         }
     }
 
+    class MapInResultDeserializer : ResponseDeserializable<Map<String, SHRemoteUser>> {
```
Comment (gennaro-safehill) [inline src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt]: For consistency with the method below: `MapDeserializer`and `ListDeserializer`

File: src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt [hunk 3]
```diff
@@ -36,6 +44,39 @@ data class SHRemoteUser(
...
+        private val mapSerializer = MapSerializer(
+            String.serializer(),
+            SHRemoteUser.serializer()
+        )
+
+        override fun deserialize(content: String): Map<String, SHRemoteUser> {
+            return Json.decodeFromString(
+                object : DeserializationStrategy<Map<String, SHRemoteUser>> {
+                    override val descriptor: SerialDescriptor =
+                        buildClassSerialDescriptor("SHRemoteUserResults") {
```
Comment (gennaro-safehill) [inline src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt]: We can name this "Map<String, SHRemoteUser>"

File: src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt [hunk 4]
```diff
@@ -36,6 +44,39 @@ data class SHRemoteUser(
         }
     }
 
+    class MapInResultDeserializer : ResponseDeserializable<Map<String, SHRemoteUser>> {
+        private val mapSerializer = MapSerializer(
+            String.serializer(),
+            SHRemoteUser.serializer()
+        )
+
+        override fun deserialize(content: String): Map<String, SHRemoteUser> {
```
Comment (gennaro-safehill) [inline src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt]: Define a typealias for user public identifier = String so we can differentiate between - for instance - asset global identifiers as strings and user public identifiers

File: src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt [hunk 5]
```diff
@@ -36,6 +44,39 @@ data class SHRemoteUser(
...
+                            map!!
+                        }
+                    }
+                },
+                content,
+            )
+        }
+    }
+
     class ListDeserializer : ResponseDeserializable<List<SHRemoteUser>> {
```
Comment (gennaro-safehill) [inline src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt]: I'm not sure these serializers should be added to the model, but they rather belong (like the other serializers) to com.safehill.kclient.api.serde
What do you think?

File: src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt (hunk 4 above)
Comment (DikenMaharjan) [reply src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt]: Are we getting user identifiers on keys in map? I thought we are getting the hashed phone numbers back as key.

Comment (gennaro-safehill) [reply src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt]: Ah, that's right. That's why we need more clarity in the types :)

File: src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt (hunk 5 above)
Comment (DikenMaharjan) [reply src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt]: We probably do not need ```SHRemoteUser.Deserializer``` and  ```SHRemoteUser.ListDeserializer``` for deserializing. We can use [fuel-kotlinx-serialization](https://fuel.gitbook.io/documentation/deserialization/fuel-kotlinx-serialization) for serialization using [KotlinX Serialization](https://github.com/Kotlin/kotlinx.serialization#gradle) in fuel.

For the ```SHRemoteUser.MapInResultDeserializer``` though, we would need a deserializer as we have some parsing to do. I do not have a strong recommendation for its location. As you said, ```com.safehill.kclient.api.serde``` is a viable choice in my opinion.

File: src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt (hunk 2 above)
Comment (DikenMaharjan) [reply src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt]: Its not really a map as a whole. The map object is contained within "result" key of the json. Do we really want to name it MapDeserializer?

File: src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt (hunk 4 above)
Comment (DikenMaharjan) [reply src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt]: What alias name do you recommend?

Comment (gennaro-safehill) [reply src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt]: Same as in the Swift codebase
```
public typealias GlobalIdentifier = String
public typealias LocalIdentifier = String
public typealias UserIdentifier = String
public typealias BearerToken = String
```

HashedPhoneNumber is currently not a type alias but it should be

File: src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt (hunk 2 above)
Comment (gennaro-safehill) [reply src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt]: Up to you but MapInResult is cryptic. Please find a better name

File: src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt (hunk 5 above)
Comment (DikenMaharjan) [reply src/main/kotlin/com/safehill/kclient/models/SHRemoteUser.kt]: Upon some thought, I dont think we would be needing ```SHRemoteUser.MapInResultDeserializer``` as well. We could have a class as shown for deserializing.
```
@Serialization
data class ContactMatchingUsers(
val result: Map<HashedPhoneNumber, SHRemoteUser>
)
```

**Review by gennaro-safehill — APPROVED** [review]
LGTM

**Pull request merged**